Transaction entry

aumayr opened the pull request:
Implementation of #421

- [x] Basic UI
- [x] Mobile UI
- [x] Helpers for UI (autocomplete, formatting, adding new lines, etc.)
- [x] Data for autocomplete
- [x] Writing the resulting transaction to the beancount file
- [x] ~~~Translations (ru, de)~~~ (shouldn't have to be part of feature-PR)
- [x] Tests
- [ ] Documentation

This is a very big feature to implement, so any help is welcome!

aumayr commented:
There are some issues/tasks still open:

- [ ] Bug: "Save and add new" adds the same transaction multiple times
- [ ] make indentation configureable
- [ ] dropdown for `*`
- [ ] currency as `input` (awesomplete)?
- [ ] multiple postings
- [ ] intelligent suggestions for account
- [ ] tags? links?
- [ ] metadata?

yagebu commented on fava/api/file.py:
```diff
@@ -32,3 +32,30 @@ def insert_line_in_file(filename, lineno, content):
         file.write(contents)
+
+
+def insert_transaction_in_file(filename, lineno, content):
```
This should really insert a `data.Transaction` and use `parser.printer` for it.

yagebu commented:
Can you rebase this and fix tests/lint? Then I could help with some of the changes

aumayr commented:
I rebased and fixed the tests/lint. 

You can display the transaction form by using the keyboard shortcut `n`. 

Please announce it (here or on Gitter) if you are working on changes, so we do not work on the same.

yagebu commented:
Seems like you didn't push the rebase, will do that.

Also the autocomplete is currently shown in the wrong position, will fix the css for that and also hide some of the labels, I find placeholders are nicer. This way we can imitate the form looks more like a transaction in the file.

aumayr commented:
> I find placeholders are nicer

Yes, this looks much nicer!

aumayr commented:
@yagebu I guess you are adding the `currencies`-`<list>` and also still work on the JavaScript (as `currencies` is now an `input`, not a `select`)?

yagebu commented:
Yes just started with improving the completions a bit.

aumayr commented:
> Yes just started with improving the completions a bit.

The algorithm(s) for suggesting the account based on the payee, narration and/or other account are an interesting (and seemingly hard) part. Also, these algorithms could be used (or extended) for implementing a UI for `bean-extract`, as suggested in https://github.com/beancount/fava/issues/421#issuecomment-263837979

yagebu commented:
> The algorithm(s) for suggesting the account based on the payee, narration and/or other account are an interesting (and seemingly hard) part. Also, these algorithms could be used (or extended) for implementing a UI for bean-extract, as suggested in #421 (comment)

Right now, I'm only working on the JS part to do fuzzy matchings (and only suggest active accounts). That already goes a long way.

yagebu commented:
I've added an input for the flag - it's "cheap" as it has a default and value and can simply be skipped. I'm against supporting links, tags and metadata - it doesn't even support costs atm (and I'm fine with that, it's meant for simple transactions IMHO).

yagebu commented on fava/templates/overlays/_transaction_form.html:
```diff
@@ -12,20 +12,21 @@
-                <input type="number" step="0.01" class="number" name="number" placeholder="Number" />
-                <input type="text" class="currency" name="currency" placeholder="Currency" list="currencies" />
+            <div class="postings">
+            {% for i in range(15) %}
```
This seems like a waste of bandwidth. [cloneNode](https://developer.mozilla.org/en/docs/Web/API/Node/cloneNode) should do the trick

aumayr commented on fava/static/sass/_overlay.scss:
```diff
@@ -65,62 +65,31 @@ $overlay-wrapper-background: rgba(0, 0, 0, .5);
-    display: block;
-    margin-top: 1rem;
-
-    &:focus {
```
The focus states are important IMHO because tabbing through the form and then pressing "enter" when the focus is on the Save is really helpful.

yagebu commented:
> These changes broke the layout for me:

Which browser are you on? Works fine in Firefox and Chrome for me

aumayr commented:
> Which browser are you on? Works fine in Firefox and Chrome for me

Chrome 54 on a Mac. Safari looks the same. I did force-refresh a couple of times, but still looks like the screenshot.

yagebu commented:
Ok, can reproduce this in Safari (not in Chrome 55). Seems like some browsers don't support `display: flex` for `fieldset` - browser bugs hurray.

lipidity commented:
Ranking autocomplete suggestions

One way of doing this would be to use [exponential decay](http://www.evanmiller.org/rank-hotness-with-newtons-law-of-cooling.html)

1. keep a dict mapping account name to some sort of tuple/dict structure of (temperature, last_date)
2. when a payee is set in the form, send an ajax query to retrieve all postings where payee = "what the user entered as payee"
3. for each posting returned in step 2, look up the account in the dict from step 1, run the cooling equation on the existing temperature (based on how much time elapsed between the last_date and the date of the current posting), increment the temperature and update the last_date to the date of this posting

I used the same approach for ledgible. This worked _very_ well.

(I used a cooling rate factor of `-1 / 30`, and temperature increment of 100)

johannesjh commented:
I tried this pull request out but I cannot add new transactions... it always produces server-side errors. I don't quite understand why these errors occur but maybe you finde the stacktraces helpful?

```
Traceback (most recent call last):
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/flask/app.py", line 1988, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/flask/app.py", line 1641, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/flask/app.py", line 1544, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/flask/app.py", line 1639, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/flask/app.py", line 1625, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/johannes/repos/fava/fava/application.py", line 373, in api_add_transaction
    g.api.validate_and_insert_transaction(transaction)
  File "/Users/johannes/repos/fava/fava/api/__init__.py", line 466, in validate_and_insert_transaction
    entries, errors = booking.book([transaction], self.options)
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/beancount/parser/booking.py", line 53, in book
    booking_methods)
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/beancount/parser/booking_simple.py", line 48, in book
    balance_errors = interpolate.balance_incomplete_postings(entry, options_map)
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/beancount/core/interpolate.py", line 498, in balance_incomplete_postings
    entry.meta = {}
AttributeError: can't set attribute
```

or this one:

```
[2016-12-05 17:28:18,215] ERROR in app: Exception on /example-beancount-file/api/add-transaction/ [PUT]
Traceback (most recent call last):
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/flask/app.py", line 1988, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/flask/app.py", line 1641, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/flask/app.py", line 1544, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/flask/app.py", line 1639, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/flask/app.py", line 1625, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/johannes/repos/fava/fava/application.py", line 373, in api_add_transaction
    g.api.validate_and_insert_transaction(transaction)
  File "/Users/johannes/repos/fava/fava/api/__init__.py", line 466, in validate_and_insert_transaction
    entries, errors = booking.book([transaction], self.options)
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/beancount/parser/booking.py", line 53, in book
    booking_methods)
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/beancount/parser/booking_simple.py", line 48, in book
    balance_errors = interpolate.balance_incomplete_postings(entry, options_map)
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/beancount/core/interpolate.py", line 480, in balance_incomplete_postings
    residual, tolerances) = get_incomplete_postings(entry, options_map)
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/beancount/core/interpolate.py", line 345, in get_incomplete_postings
    tolerances = infer_tolerances(postings, options_map)
  File "/Users/johannes/repos/fava/venv/lib/python3.5/site-packages/beancount/core/interpolate.py", line 212, in infer_tolerances
    expo = units.number.as_tuple().exponent
AttributeError: type object 'MISSING' has no attribute 'as_tuple'
```

tested locally on commit 8c95a58ac6eaf4cbdd7ce73f0a916e4a642ecc1b, by running fava against my own ledger, as well as against a file generated using bean-example.

yagebu commented:
Thank you @lipidity for the suggestion! I'd be for implementing the enhanced completion in a later PR, this one already contains a lot of changes.

@johannesjh: I guess the validation only works when using the "FULL" booking algorithm in Beancount. This part of the PR is still incomplete, will hopefully have time later this week to finish it.

johannesjh commented:
1. Regarding the server-side errors:
I tried it again, this time using the "FULL" booking algorithm. Now it works, thank you!

    Note, for further reference: 
    Adding this line to the beancount file will solve the problem:
    `option "experiment_booking_algorithm" "FULL"`

2. Regarding mobile use:
I also tried adding a new entry using my mobile phone because that is one usage scenario that I'm interested in. This did not work because I could not find a way to trigger the keyboard shortcut `n`. Could you add a menu entry?

lipidity commented:
I had the same question as @johannesjh about opening the add transaction form on a mobile device.

Also, just curious - why an overlay instead of a dedicated page?

yagebu commented:
I've removed some of the validation - I don't think we really need to check that the transaction balances - but added checks for the accounts. I've also added a button to the sidebar to open the form.

> Also, just curious - why an overlay instead of a dedicated page?

Being able to stay on the same page when adding a transaction is nice, a separate page would also be overkill for such a small form IMHO.

aumayr commented:
A few notes/issues after testing:

- This feature is awesome. It saves a lot of time, and with a few tweaks could be even faster.
- When selecting whitespace in the editor, the red whitespace-background is shown, not the selection-background-color
- When inserting transactions, unnecessary whitespace is added at the end of postings and has to be removed manually in the editor
- The "+"-button and keyboard shortcut are not working when on the Editor page
- When adding a third posting row and more, awesomplete is not initialized for those rows
- What would be awesome: I always enter transactions after some days have passed. "Save and add new" is a real help here. What would make it even faster would be that the date does not get resettet to today, but stays the same as before. 

The formatting of the inserted transactions is not ideal for me. I know that this is a matter of style and preference, as beancount allows for some variations. Here an example what get's inserted:

```
2016-12-06 * "Testnarration 1"
  Expenses:Some                              6 EUR                                              
  Assets:Some                                <class 'beancount.core.number.MISSING'> <class 'beancount.core.number.MISSING'>

2016-12-07 * "Testpayee" | "Testnarration2"
  Expenses:Some                              6 EUR                                              
  Assets:Some                                -6.00 EUR
```

- I always use two decimal places for EUR. There is no way to configure how many should be added/shown/inserted.
- I always omit the value and currency for the second/last posting (get's calculated by beancount automatically). When omitting the value and currency in the form, `<class 'beancount.core.number.MISSING'> <class 'beancount.core.number.MISSING'>` get's inserted. 
- I always omit the "`|`" between payee and narration. Should be configureable as well IMHO.
- I never omit the `""` when there is no payee. So the first line would be `2016-12-06 * "" "Testnarration 1"` in my files. (I can live with omitting this, but just want to note it here.)

corani commented:
I believe the "|" between payee and narration is deprecated already.

As for decimal places, ideally this should be derived from the context (I'm
a bit vague on the exact term), so it'll automatically match what you're
using in the source.

I too anyways use two decimals, and omit the amount on one of the postings
(unless it's an especially complicated entry, then i enter it to ensure I
didn't make any mistakes)

On Thu, Dec 8, 2016, 17:58 Dominik Aumayr <notifications@github.com> wrote:

> A few notes/issues after testing:
>
>    - This feature is awesome. It saves a lot of time, and with a few
>    tweaks could be even faster.
>    - When selecting whitespace in the editor, the red
>    whitespace-background is shown, not the selection-background-color
>    - When inserting transactions, unnecessary whitespace is added at the
>    end of postings and has to be removed manually in the editor
>    - The "+"-button and keyboard shortcut are not working when on the
>    Editor page
>    - When adding a third posting row and more, awesomplete is not
>    initialized for those rows
>    - What would be awesome: I always enter transactions after some days
>    have passed. "Save and add new" is a real help here. What would make it
>    even faster would be that the date does not get resettet to today, but
>    stays the same as before.
>
> The formatting of the inserted transactions is not ideal for me. I know
> that this is a matter of style and preference, as beancount allows for some
> variations. Here an example what get's inserted:
>
> 2016-12-06 * "Testnarration 1"
>   Expenses:Some                              6 EUR
>   Assets:Some                                <class 'beancount.core.number.MISSING'> <class 'beancount.core.number.MISSING'>
>
> 2016-12-07 * "Testpayee" | "Testnarration2"
>   Expenses:Some                              6 EUR
>   Assets:Some                                -6.00 EUR
>
>
>    - I always use two decimal places for EUR. There is no way to
>    configure how many should be added/shown/inserted.
>    - I always omit the value and currency for the second/last posting
>    (get's calculated by beancount automatically). When omitting the value and
>    currency in the form, <class 'beancount.core.number.MISSING'> <class
>    'beancount.core.number.MISSING'> get's inserted.
>    - I always omit the "|" between payee and narration. Should be
>    configureable as well IMHO.
>    - I never omit the "" when there is no payee. So the first line would
>    be 2016-12-06 * "" "Testnarration 1" in my files. (I can live with
>    omitting this, but just want to note it here.)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/beancount/fava/pull/426#issuecomment-265700538>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAdWB4N6B8XF82vUovh83e5rM6eT2ELOks5rF9TTgaJpZM4K9Ivk>
> .
>
-- 

Best regards,
Daniël Bos

Your government is reading your email. Slow them down with encryption.

My public key: http://goo.gl/gms497 (4096 bit RSA, id EF2D5D91)
Fingerprint  : D8D0 9FBE F075 F709 7B52  2F73 326C 2123 EF2D 5D91

blais commented:
I haven't been able to follow this thread in its entirety, but let me
comment on a few things that might help:

- "|" is officially deprecated now.

- Inferring the number of decimal places required for a number can be done
by using a DisplayContext object, which is initialized from the input file
to reflect the usage in that file (does some basic statistics on numbers
used in the input). Not sure if it's what  you might be looking for here.

- beancount.core.number.MISSING objects are the result of running just the
parser phase on some input with missing values left to be interpolated. The
booking code always resolves these away, plugins never seen them. It's a
pipeline: input -> parser -> booking -> plugins -> resulting entries. Not
sure what's going on in this particular example. If you're running the
parser bit and then outputting using the printer, the printer should be
enhanced to make it possible to print incomplete entries with MISSING
objects.

Should I bake a new release? Doing so would force everyone to use the new
booking code, which is vastly better, but still has a remaining flaw (which
I want to fix soon): https://bitbucket.org/blais/beancount/issues/145




On Thu, Dec 8, 2016 at 4:58 AM, Dominik Aumayr <notifications@github.com>
wrote:

> A few notes/issues after testing:
>
>    - This feature is awesome. It saves a lot of time, and with a few
>    tweaks could be even faster.
>    - When selecting whitespace in the editor, the red
>    whitespace-background is shown, not the selection-background-color
>    - When inserting transactions, unnecessary whitespace is added at the
>    end of postings and has to be removed manually in the editor
>    - The "+"-button and keyboard shortcut are not working when on the
>    Editor page
>    - When adding a third posting row and more, awesomplete is not
>    initialized for those rows
>    - What would be awesome: I always enter transactions after some days
>    have passed. "Save and add new" is a real help here. What would make it
>    even faster would be that the date does not get resettet to today, but
>    stays the same as before.
>
> The formatting of the inserted transactions is not ideal for me. I know
> that this is a matter of style and preference, as beancount allows for some
> variations. Here an example what get's inserted:
>
> 2016-12-06 * "Testnarration 1"
>   Expenses:Some                              6 EUR
>   Assets:Some                                <class 'beancount.core.number.MISSING'> <class 'beancount.core.number.MISSING'>
>
> 2016-12-07 * "Testpayee" | "Testnarration2"
>   Expenses:Some                              6 EUR
>   Assets:Some                                -6.00 EUR
>
>
>    - I always use two decimal places for EUR. There is no way to
>    configure how many should be added/shown/inserted.
>    - I always omit the value and currency for the second/last posting
>    (get's calculated by beancount automatically). When omitting the value and
>    currency in the form, <class 'beancount.core.number.MISSING'> <class
>    'beancount.core.number.MISSING'> get's inserted.
>    - I always omit the "|" between payee and narration. Should be
>    configureable as well IMHO.
>    - I never omit the "" when there is no payee. So the first line would
>    be 2016-12-06 * "" "Testnarration 1" in my files. (I can live with
>    omitting this, but just want to note it here.)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/beancount/fava/pull/426#issuecomment-265700538>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAUgkwXbSKmamvMcalQhKSEO5hxHlfThks5rF9TTgaJpZM4K9Ivk>
> .
>

xentac commented:
I already use the dev version consistently, so don't hold back the release
on my account.  I'm sure there are some other people who use autosale
though who may be tracking their transactions the same way I am.

On Thu, Dec 8, 2016 at 8:15 AM Martin Blais <notifications@github.com>
wrote:

> I haven't been able to follow this thread in its entirety, but let me
> comment on a few things that might help:
>
> - "|" is officially deprecated now.
>
> - Inferring the number of decimal places required for a number can be done
> by using a DisplayContext object, which is initialized from the input file
> to reflect the usage in that file (does some basic statistics on numbers
> used in the input). Not sure if it's what you might be looking for here.
>
> - beancount.core.number.MISSING objects are the result of running just the
> parser phase on some input with missing values left to be interpolated. The
> booking code always resolves these away, plugins never seen them. It's a
> pipeline: input -> parser -> booking -> plugins -> resulting entries. Not
> sure what's going on in this particular example. If you're running the
> parser bit and then outputting using the printer, the printer should be
> enhanced to make it possible to print incomplete entries with MISSING
> objects.
>
> Should I bake a new release? Doing so would force everyone to use the new
> booking code, which is vastly better, but still has a remaining flaw (which
> I want to fix soon): https://bitbucket.org/blais/beancount/issues/145
>
>
>
>
> On Thu, Dec 8, 2016 at 4:58 AM, Dominik Aumayr <notifications@github.com>
> wrote:
>
> > A few notes/issues after testing:
> >
> > - This feature is awesome. It saves a lot of time, and with a few
> > tweaks could be even faster.
> > - When selecting whitespace in the editor, the red
> > whitespace-background is shown, not the selection-background-color
> > - When inserting transactions, unnecessary whitespace is added at the
> > end of postings and has to be removed manually in the editor
> > - The "+"-button and keyboard shortcut are not working when on the
> > Editor page
> > - When adding a third posting row and more, awesomplete is not
> > initialized for those rows
> > - What would be awesome: I always enter transactions after some days
> > have passed. "Save and add new" is a real help here. What would make it
> > even faster would be that the date does not get resettet to today, but
> > stays the same as before.
> >
> > The formatting of the inserted transactions is not ideal for me. I know
> > that this is a matter of style and preference, as beancount allows for
> some
> > variations. Here an example what get's inserted:
> >
> > 2016-12-06 * "Testnarration 1"
> > Expenses:Some 6 EUR
> > Assets:Some <class 'beancount.core.number.MISSING'> <class
> 'beancount.core.number.MISSING'>
> >
> > 2016-12-07 * "Testpayee" | "Testnarration2"
> > Expenses:Some 6 EUR
> > Assets:Some -6.00 EUR
> >
> >
> > - I always use two decimal places for EUR. There is no way to
> > configure how many should be added/shown/inserted.
> > - I always omit the value and currency for the second/last posting
> > (get's calculated by beancount automatically). When omitting the value
> and
> > currency in the form, <class 'beancount.core.number.MISSING'> <class
> > 'beancount.core.number.MISSING'> get's inserted.
> > - I always omit the "|" between payee and narration. Should be
> > configureable as well IMHO.
> > - I never omit the "" when there is no payee. So the first line would
> > be 2016-12-06 * "" "Testnarration 1" in my files. (I can live with
> > omitting this, but just want to note it here.)
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/beancount/fava/pull/426#issuecomment-265700538>, or
> mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAUgkwXbSKmamvMcalQhKSEO5hxHlfThks5rF9TTgaJpZM4K9Ivk
> >
> > .
> >
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/beancount/fava/pull/426#issuecomment-265780814>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAHDYeUYsO6NF_BV4Gc33lHKCrhS_MLks5rGC0igaJpZM4K9Ivk>
> .
>

aumayr commented:
> Should I bake a new release? Doing so would force everyone to use the new
booking code, which is vastly better, but still has a remaining flaw (which
I want to fix soon): https://bitbucket.org/blais/beancount/issues/145

For development we use the current code from bitbucket. I guess when this feature is complete we release a new version, for which we would need a new version/release of beancount as well (for people installing via pip). We will ping you once this is approaching.

aumayr commented:
The currency-input should change it's width for the last row, so it makes space for the "+" and the number inputs are all the same width and neatly vertically aligned.

Currently it looks like this:

![screenshot](https://cloud.githubusercontent.com/assets/5135450/21073421/bc95d5bc-bedd-11e6-98cb-2635edc60af0.png)

@yagebu Do you know how to achieve that?

aumayr commented on fava/templates/overlays/_transaction_form.html:
```diff
@@ -13,8 +13,8 @@
-            {% for i in range(2) %}
-                <div class="fieldset posting">
+            {% for i in range(3) %}
+                <div class="fieldset posting{% if i == 0 %} template{% endif %}">
```
Previous to this change, if you add new posting rows with the "+" the inputs in those rows were not awesomplete-initialized, so there was no auto-complete of them. 

Background:
When cloning the `.posting`-row after pressing "+", the nodes generated by awesomplete were cloned, and the auto-completion did not "carry over" to those nodes, as the `list="..."`-property got removed by awesomplete.

yagebu replied:
Makes sense, thanks for the explanation!

yagebu commented:
This now uses a (very stupid) handwritten printer that should address most of the concerns above.

> I always use two decimal places for EUR. There is no way to configure how many should be added/shown/inserted.

Currently it simply uses whatever the user inputs -- which seems the right thing to do to me.

> Currently it looks like this:

That's the look I was going for, I didn't like the tiny currency input. Changing the width's in the conditional `last-child` code would be the way to change it.

yagebu commented:
> That's the look I was going for, I didn't like the tiny currency input. Changing the width's in the conditional last-child code would be the way to change it.

This should be understood as a "feel free to change it, I don't care much" ;)

yagebu commented:
> The "+"-button and keyboard shortcut are not working when on the Editor page

Can't reproduce. (Of course when the editor is focused the keyboard shortcut doesn't work, but otherwise the shortcut and the button work for me)

aumayr commented:
>> The "+"-button and keyboard shortcut are not working when on the Editor page

>Can't reproduce. (Of course when the editor is focused the keyboard shortcut doesn't work, but otherwise the shortcut and the button work for me)

It works now - sorry for the confusion!

aumayr commented:
There are just two tasks open:

- [ ] When clicking into the results list from awesomplete and the result clicked on is outside the bounds of the overlay, the overlay is dismissed. I tried to fix this, but somehow magically a MouseEvent appears when calling `input.close()` on the Awesomplete-element. So basically I don't know how to fix this. @yagebu Do you have any clue why this happens?
- [ ] Fix the styling for mobile browsers.

After this is complete, Documentation and Translation have to be added as well.

yagebu commented:
>  When clicking into the results list from awesomplete and the result clicked on is outside the bounds of the overlay, the overlay is dismissed. I tried to fix this, but somehow magically a MouseEvent appears when calling input.close() on the Awesomplete-element. So basically I don't know how to fix this. @yagebu Do you have any clue why this happens?

Awesomplete listened to `mousedown` and we to `click` (which is basically `mousedown`+`mouseup`) which would fire after the list had been closed already.

yagebu commented on fava/static/sass/_base.scss:
```diff
@@ -290,10 +290,33 @@ header {
 
+.add-transaction-button {
+  background-color: $color-sidebar-background;
+  color: $color-journaltable-header;
```
this is definitely not a variable that should be used here

yagebu commented on fava/static/sass/_base.scss:
```diff
@@ -584,6 +608,10 @@ td {
       margin-left: 0;
+
+      .add-transaction-button {
+        left: 160px;
```
try not to use random constants like this one, we've got variables for this.

yagebu commented on fava/static/sass/_base.scss:
```diff
@@ -592,6 +620,18 @@ td {
+    height: 41px;
+    left: 0;
+    position: fixed;
+    top: 95px;
```
ditto

aumayr commented:
@fokusov Could you provide the Russian translation strings?

fokusov commented:
@aumayr  Yes, I'm working on it
> @fokusov Could you provide the Russian translation strings?

fokusov commented:
@aumayr I can't find no new translated strings with `make babel-extract` in updated templates :(

aumayr commented:
@fokusov Oh, there seem to be no new strings to translate after all 😃 I thought I had translated new parts for German before and the Russian counterparts were missing, but that does not seem to be the case. 

Anyhow, thanks for the help!
